### DISCUSSION TITLE
Apakah mungkin untuk menghindari skema URL dengan akhiran menyerupai dotfile?

### DISCUSSION THREAD

**Discussion by salimromadhon** [mecha-cms/mecha#296] [Q&A] [answered]
Halo semuanya.

Saya menemukan kasus di mana skema URL seperti `https://example.com/panel/get/.state` menyebabkan masalah di penyedia layanan hosting saya. Setelah saya konfirmasi ke pihak hosting, alasannya cukup masuk akal. URL tersebut menyerupai alur berkas tersembunyi yang semestinya tidak bisa diakses oleh publik. Mereka menyarankan saya untuk mengubah skema URL-nya alih-alih mengabulkan permintaan saya untuk mengizinkan skema URL tersebut.

Kira-kira apakah ada opsi untuk mengubah skema URL ini di Mecha?

Terima kasih.

**Comment by taufik-nurrohman** [comment, marked as answer]
Untuk saat ini belum mas, itu sepertinya ada dari pengaturan server inti, untuk proteksi. Harusnya bagusnya itu hanya berlaku untuk berkas yang eksis, bukan untuk rute buatan yang tidak mengarah ke berkas apa-apa.

Saya bisa mempertimbangkan untuk menambahkan pengaturan prefiks kustom pada kasus ini, tapi sebelum itu Saya mau cari tahu dulu sekiranya ada solusi yang bisa dicoba melalui berkas `.htaccess`.

Beberapa kasus juga terjadi pada server Windows yang tidak mau menerima karakter `:` pada URL, padahal Wikipedia pakai itu di beberapa tempat dan sepertinya Wikipedia tidak peduli.

![Fix Potential Bug on Windows Server](https://github.com/user-attachments/assets/67f624bc-fc66-4a9d-bdcc-7b1aceaf3794)

**Reply by salimromadhon** [reply]
> Harusnya bagusnya itu hanya berlaku untuk berkas yang eksis, bukan untuk rute buatan yang tidak mengarah ke berkas apa-apa.

Sepakat Mas. Sayangnya pihak hosting tidak peduli...

Saya sempat coba cari solusinya pakai `.htaccess`, tapi tidak berhasil.

> Beberapa kasus juga terjadi pada server Windows yang tidak mau menerima karakter : pada URL, padahal Wikipedia pakai itu di beberapa tempat dan sepertinya Wikipedia tidak peduli.

Oh iya betul, soal karakter titik dua, seingat saya pernah juga saya alami. Kalau tidak salah, Windows menerima titik koma sebagai gantinya, meskipun aneh.

Terlepas dari masalah di atas, sebenarnya saya suka dengan skema URL sekarang. Sedih juga jadi karena berimbas ke portability Mecha.

Terima kasih Mas, sudah berkenan mempertimbangkan. Tidak sabar untuk kabar selanjutnya. :slightly_smiling_face:

**Reply by yuliakbard** [reply]
ini pernah sempat ane sampaikan ke mas taufik perihal ini. kalau gk salah ini kejadian waktu pakai hosting utk url yg ada \.state gk bisa akses dan tiba tiba bisa akses seperti biasa kembali sehabis utak atik bagian hosting, cuma gk tau bagian mana yg ane setting sampai bisa nge akses url dotfile.

**Reply by taufik-nurrohman** [reply]
Kalau simbol `.` diganti jadi `%2E` ngaruh nggak mas @salimromadhon ?

~~~ txt
/panel/get/%2Estate
~~~

Seenggaknya bisa nampilin laman 404 gitu.

**Reply by yuliakbard** [reply]
apa gk bisa ditiadakan simbol `.` dari url nya kang, biar lebih clean

**Reply by taufik-nurrohman** [reply]
Bisa tapi itu sengaja Saya pakai format jalur seperti itu untuk menunjukkan bahwa rute tersebut tidak mengarah ke berkas/folder yang eksis (rute kustom). Kalau misal Saya standarkan jalurnya menjadi `/panel/get/state/1` seharusnya akan lebih ideal untuk menampilkan isi folder `.\lot\state`.

**Reply by salimromadhon** [reply]
```
/panel/get/%2Estate
```

Sama saja Mas, masih error 520.

Kalau diganti `+state` cakep engga ya? Kaya Google+ dulu.

```
/panel/get/+state
```

**Reply by taufik-nurrohman** [reply]
Niatnya mau tak bikin _configurable_ aja si, cuma sifatnya kaya _patch_ aja nanti, jadi tidak masuk ke dokumentasi.

Itu kemungkinan karena ada pengaturan [`<LocationMatch "^/[.]">`](https://httpd.apache.org/docs/current/mod/core.html#locationmatch) di `httpd.conf` yang cuma bisa diakses oleh _super user_.

**Reply by salimromadhon** [reply]
Bisa jadi Mas.

Memang secara visual lebih oke pake titik daripada yang lain. Paling tidak, yang nasibnya kurang beruntung dengan masalah tersebut bisa tetap pakai Mecha.

**Comment by taufik-nurrohman** [comment]
Untuk realisasinya secara permanen sebagai rilis baru sepertinya agak berat mas, karena dalam rute penyunting berkas itu dia mengikuti nama berkas yang asli. Jadi sebagai contoh, ada sebuah berkas `.htaccess` di dalam folder ekstensi, dan itu seharusnya bisa disunting melalui Panel. Kalau pakai pola tautan yang berbeda dari nama berkas asli kan tentunya akan cukup mengagetkan pengguna (membuat bingung).

Tapi ini Saya ada solusi simpel untuk membuat Mecha bisa menganggap awalan `/+` sebagai `/.`. Karena toh _server_ menolak pola tautan dengan awalan `/.` sebelum mencapai eksekusi kode PHP. Jadi kalau nilai rute yang bisa diterima oleh _server_ kemudian diubah menjadi nilai rute yang tidak diterima oleh _server_ sebelum rute tersebut diproses oleh PHP, maka masalah akan teratasi. Caranya buka berkas `.\engine\fire.php` kemudian cari baris instansiasi objek `URL` ini:

``` php
lot('url', $url = new URL($protocol . $host . $path . $query . $hash));
```

Kemudian sebelum baris di atas, tambahkan kode ini:

``` php
$path = strtr($path, ['/+' => '/.']);
```

Simpan. Sekarang seharusnya mas bisa mengubah simbol `.` menjadi `+` pada tautan apabila _server_ menolak tautan tersebut.

Untuk mengubah seluruh tautan di dalam situs web secara otomatis sesuai dengan spesifikasi yang baru bisa dilakukan dengan menambahkan kait ini:

``` php
Hook::set('content', function ($content) {
    if (!$content) {
        return $content;
    }
    $pattern = x(lot('url') . '/') . '[^"\'\s]+';
    return preg_replace_callback('/' . $pattern . '/', static function ($m) {
        return strtr($m[0], ['/.' => '/+']);
    }, $content);
});
```

Lihat juga: <https://superuser.com/a/1645935>

**Reply by salimromadhon** [reply]
Saya berubah pikiran untuk mengganti `+` dengan `~`. Nampaknya juga berhasil.

Terima kasih, Mas. :)
